File: src/test/java/fr/marstech/mtlinkspray/controller/api/ShortenerApiControllerTest.java
package fr.marstech.mtlinkspray.controller.api;

import fr.marstech.mtlinkspray.service.ShortenerServiceImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.net.URI;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
class ShortenerApiControllerTest {

  @Autowired private MockMvc mockMvc;

  @MockitoBean private ShortenerServiceImpl shortenerServiceImpl;

  @BeforeEach
  void setUp() throws Exception {
    when(shortenerServiceImpl.shorten(anyString(), any()))
        .thenReturn(URI.create("https://short.url").toURL().toString());
  }

  @Test
  void getShort() throws Exception {
    mockMvc
        .perform(get("/api/url-shortener/shorten").param("url", "https://www.example.com"))
        .andExpect(status().isOk())
        .andExpect(content().string("https://short.url"));
  }
}

File: src/test/kotlin/fr/marstech/mtlinkspray/controller/api/ShortenerApiControllerTest.kt
package fr.marstech.mtlinkspray.controller.api

import fr.marstech.mtlinkspray.service.ShortenerService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.content
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders

class ShortenerApiControllerTest {

    private lateinit var mockMvc: MockMvc
    private val shortenerService: ShortenerService = mock()

    @BeforeEach
    fun setup() {
        val validator = org.springframework.validation.beanvalidation.LocalValidatorFactoryBean()
        validator.afterPropertiesSet()
        
        mockMvc = MockMvcBuilders.standaloneSetup(ShortenerApiController(shortenerService))
            .setControllerAdvice(fr.marstech.mtlinkspray.controller.commons.GlobalRestExceptionHandler())
            .setValidator(validator)
            .build()
    }

    @Test
    fun shouldGetShortenedUrl() {
        val inputUrl = "https://example.com"
        val shortenedUrl = "https://short.ly/abc123"

        whenever(shortenerService.shorten(any<String>(), any())).thenReturn(shortenedUrl)

        mockMvc.perform(get("/api/url-shortener/shorten").param("url", inputUrl))
            .andExpect(status().isOk)
            .andExpect(content().string(shortenedUrl))
    }

    @Test
    fun shouldReturn400ForMissingUrl() {
        mockMvc.perform(get("/api/url-shortener/shorten"))
            .andExpect(status().isBadRequest)
    }

    @Test
    fun shouldReturn400ForBlankUrl() {
        mockMvc.perform(get("/api/url-shortener/shorten").param("url", ""))
            .andExpect(status().isBadRequest)
    }

    @Test
    fun shouldReturn400ForWhitespaceUrl() {
        mockMvc.perform(get("/api/url-shortener/shorten").param("url", "   "))
            .andExpect(status().isBadRequest)
    }

    @Test
    fun shouldHandleServiceException() {
        whenever(shortenerService.shorten(any<String>(), any()))
            .thenThrow(IllegalArgumentException("Service error"))

        mockMvc.perform(get("/api/url-shortener/shorten").param("url", "https://example.com"))
            .andExpect(status().isBadRequest)
    }
}